chapter/background.md

```diff
@@ -0,0 +1,607 @@
+
+### JSON-RPC
+
+JSON-RPC (v2) [@json-rpc] is a JSON based lightweight transport independent remote procedure call [@rpc] protocol used by the LSP to communicate between a language server and a client.
+
+RPC is a well known paradigm that allows clients to virtually invoke a method at a connected process.
+The caller sends a well-defined message to a connected process which executes a procedure associated with the request, taking into account any transmitted arguments.
+Upon invoking a remote procedure, the client suspends the execution of its environment while it awaits an answer of the server, corresponding to a classical local procedure return.
+
+JSON-RPC is a JSON based implementation of RPC.
+The protocol specifies the format and meaning of the messages exchanged using the protocol, in particular *Request* and *Notification* and *Result* messages.
+Since the protocol is based on JSON, is possible to be used through any text based communication channel which includes streams or network sockets.
+In fact, JSON-RPC only specifies a message protocol, leaving the choice of the transport method to the application. 
+
+All messages refer to a `method` and a set of `parameters`.
+Servers are expected to perform the same procedure associated with the requested `method` at any time.
+In return, clients have to follow the calling conventions for the requested method.
+Typically, messages are synchronous, i.e., the client awaits a result associated to the method and its parameters before it continues its execution of the calling environment.
+Hence, requests are marked with an `id` which is included in the result/error message necessarily sent by the server.
+If the client does not require a response, it can omit the `id` field.
+The message is then interpreted as a *notification*, which servers cannot respond to.
+
+Part of the JSON-RPC specification is the ability for clients to batch requests and send a list of request or notification objects.
+In this case, the server should respond with a list of results matching each request, yet is free to process requests concurrently.
+
+### Commands and Notifications
+
+The LSP builds on top of the JSON-RPC protocol described in the previous subsection.
+In total the LSP defines 33 [@lsp] "language features", i.e., source code related capabilities.
+In addition, the LSP specifies different capabilities to the server to control the editor.
+For instance, servers may instruct clients to show notifications or progress bars or open documents.
+Similarly, the client has multiple ways of notifying the server of file changes, including renaming r deletion of files.
+
+This thesis aims to implement a fundamental set of capabilities.
```

Comment:
not a sentence for here. Consider that this section is self-contained, it should be readable independently from the rest of the thesis.